### DISCUSSION TITLE
Setting up perl and dependencies on Mac

### DISCUSSION THREAD

**Discussion by stefankinell** [karttapullautin/karttapullautin#3] [General]
You mentioned that you had the script working on a Mac, right? I am trying to get that working. But struggle to get the dependencies installed correct.

I have perl, installed GD / libgd via homebrew. But when I run the script I get errors on gd.pm not found in @INC. I see it installed under homebrew but can not understand what to change.

Is this how you made it work as well @rphlo ?

**Comment by rphlo** [comment]
you have to run `sudo cpan force install GD`

**Comment by rphlo** [comment]
Same goes for other dependencies

`sudo cpan force install POSIX`
`sudo cpan force install Config::Tiny`

**Comment by stefankinell** [comment]
Thanks
I got those to work after installing some other stuff as well (e.g. pkg-config).

One more dependency needed to be installed on my machine:

`sudo cpan force install Geo::ShapeFile`

What version of las2txt are you using? First I forgot it, got an error that las2txt.exe was missing. When I took that file (which is a windows-file) I got error messages about that 

```
Preparing input file
.....sh: las2txt: command not found
```

**Reply by rphlo** [reply]
You'll need to compile las2txt from source...
I'll write you the command to run to do that later...

**Reply by rphlo** [reply]
Here is how to compile las2txt
```
git clone https://github.com/LAStools/LAStools
cd LAStools
make
cp bin/las2txt /usr/local/bin/
```

**Reply by stefankinell** [reply]
Thanks, that works great!

**Comment by stefankinell** [comment]
Cool stuff! I just managed to get the perl script working on a .xyz-file I prepared in windows. So that part works.

Although it seems the curves comes out a bit different with this script comparing with the stand alone windows pullauta.exe. Is this intentionally or should I check some more? I am pretty sure I ran it with the same .ini-file settings but I can verify that.

**Reply by rphlo** [reply]
Can you link the 2 output files to see the differences?

**Reply by rphlo** [reply]
Also did you notice the more important speed improvement?

**Reply by stefankinell** [reply]
> Can you link the 2 output files to see the differences?

I did some more tests. I'll start a separate topic so we keep this one related to questions about setting up for a Mac.

**Comment by stefankinell** [comment]
I did not manage to see any differences on a Mac from a folder where I have only the perl-script and a folder where I have all the files from the code directory. But honestly I do not understand if that is because both directories are using your section with new code. The output looks very similar if not identical in the terminal window.

Comparing with my PC I get a huge speed increase. I finish a file in between 11 and 13 minutes instead of 22 on my PC. This can of course also depend on speed of computer where the PC has a Core i9-12900 (2,4 to 5,1 Ghz) and the Mac is a M1 3,2 Ghz.

**Comment by stefankinell** [comment]
Did you get batch mode to work? When I run 
`perl pullauta`
without the input file of a laz-file I can not get the script to process the files from an "in"-folder. I have tried both relative and absolute folder names in the config file.

**Reply by stefankinell** [reply]
I found some stuff in the script that needs to be updated. Like changing "dir" to "ls". I almost get it to work. Stuck on how to replace 

`system("gnome-terminal -- perl pullauta startthread $i");`
on line 1610 to something Mac-friendly.

**Reply by rphlo** [reply]
Yes, I changed the behaviour for batch mode in latest release, the batch mode should now work, instead of opening new terminals it uses multiple CPU threads.
It seems to work however I haven't fully tested a full flow.

**Reply by stefankinell** [reply]
I saw the changes for batch mode so I tried it today, works well the first runs I have done. Now I am going to play around with the optimal parallel processes and see what settings that works fastest for me. I run on a MacBook Pro with M1 Max 64 Gb. That processor is single thread/core so I'll see if it makes and difference setting the number of processes higher than 10 parallell processes. 

Well done so far and thanks!

**Comment by stefankinell** [comment]
I did some comparisons with batch speed, seeing what is the optimal number of parallel processes for my Mac. Perhaps it can help others to think of how to optimize their performance. 

As mentioned above, I ran on a MacBook Pro with M1 Max 64 Gb. It has 10 cores, 8 performance, 2 efficiency cores. Each core can run one thread.

Not so surprisingly, you will not get a faster output if you use more than 10 parallel processes with the script. I tried a batch of 100 laz-files in the area around Kiruna (https://kartor.gokartor.se/#10/7536408.0000/727256.0000) with quite a lot of cliffs and contours. I ran tests with 5 process, 10, 25 and 100. 

10, 25 and 100 all ended after 2 h 43 minutes (with a difference of 2 minutes fastest to slowest)
5 process after 4 h 35.

As a comparison, the 10 parallell process would indicate an average of 16 minutes 12 seconds. When I run a single laz-file, using the most complex (biggest file) it takes 14 minutes. the speed difference can be due to 8 high performance cores vs efficency cores also used in the batch.

Anyhow - Impressive speed. Very happy with the performance on Mac!

**Comment by stefankinell** [comment]
Second comment - speed difference on perl vs perl/rust versions.

Using single file to test I can shave off 40% of the time with the first part translated to Rust. Very impressive. I hope you can have energy and time to do the same for cliffs and vegetation parts of the script, which is where I still see time being consumed. Again, very promising and impressed. Big thanks!

**Comment by rphlo** [comment]
The code is now free of Perl code
